rebalance/rebalance.py

```diff
@@ -168,16 +174,40 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi = Non
     description = "%s to %s" % (outgoing_scid, incoming_scid)
     invoice = plugin.rpc.invoice(msatoshi, label, description, retry_for + 60)
     payment_hash = invoice['payment_hash']
-    success_msg = ""
-    try:
-        excludes = []
-        # excude all own channels to prevent unwanted shortcuts [out,mid,in]
-        mychannels = plugin.rpc.listchannels(source=my_node_id)['channels']
-        for channel in mychannels:
-            excludes += [channel['short_channel_id'] + '/0', channel['short_channel_id'] + '/1']
 
+    rpc_result = None
+    count = 0
+    excludes = [my_node_id]  # excude all own channels to prevent shortcuts
+    nodes = {}               # here we store erring node counts
+    maxhops = 1              # retry higher when getroute fails
+    msat_factor_max = 4
+
+    try:
         while int(time.time()) - start_ts < retry_for and not plugin.rebalance_stop:
-            r = plugin.rpc.getroute(incoming_node_id, msatoshi, riskfactor=1, cltv=9, fromid=outgoing_node_id, exclude=excludes)
+            count += 1
+            try:
+                r = plugin.rpc.getroute(incoming_node_id,
+                                        fromid=outgoing_node_id,
+                                        exclude=excludes,
+                                        msatoshi=msatoshi * min(maxhops, msat_factor_max),
```

Comment:
So `maxhops` can be between 1 and 5? It can be a reasonable choice, reducing lots of unnecessary route searches.

And this `min(maxhops, msat_factor_max)` multiplier is:
```
maxhops=1 -> 1
maxhops=2 -> 2
maxhops=3 -> 3
maxhops=4 -> 2
maxhops=5 -> 2
```
Is this what we want? (I don't know it is good or bad, just asking. Maybe I need to sleep one to think about this. :) )

Reply:
yes, good night. Just one question, maybe you read this tomorrow: I just tried to rebalance our channel so that your node has less balance on you side, meaning I used our channel as incoming_scid. Funnily your node gave me `WIRE_TEMPORARY_CHANNEL_FAILURE` even thou you have enough liqudity to push to me:

```
2021-05-28T22:38:26.322Z DEBUG   plugin-rebalance.py: count:1 maxhops:1 RpcError: RPC call failed: method: 
waitsendpay, payload: {'payment_hash': 'd3b07b31919b580272942d8672f81fb55ffd3d5de92015d684f951eb514e8467', 
'timeout': 600}, error: {'code': 204, 'message': 'failed: WIRE_TEMPORARY_CHANNEL_FAILURE (reply from remote)', 
'data': {'id': 224, 'payment_hash': 'd3b07b31919b580272942d8672f81fb55ffd3d5de92015d684f951eb514e8467', 
'destination': '02a4aae2e021b0785f2f2a7c9dc9089361995085298598992bf69f11302a5dbd2c', 'msatoshi': 1000000000, 
'amount_msat': 1000000000msat, 'msatoshi_sent': 1000411495, 'amount_sent_msat': 1000411495msat, 'created_at': 
1622241503, 'status': 'pending', 'erring_index': 2, 'failcode': 4103, 'failcodename': 
'WIRE_TEMPORARY_CHANNEL_FAILURE', 'erring_node': 
'03d67f36c4f81789e2fe425028bacc96b199813eae426c517f589a45f1136c1fe5', 'erring_channel': '685132x1087x0', 
'erring_direction': 1, 'raw_message': '1...''}}
```

Are there other reasons (aside imbalanced channel) that a forward can fail with `WIRE_TEMPORARY_CHANNEL_FAILURE` ??

Reply:
And also ACINQ node seems to give me `WIRE_CHANNEL_DISABLED` , ever had this?

Reply:
I've checked my logs and I've found this:
```
2021-05-28T22:38:25.203Z INFO    02a4aae2e021b0785f2f2a7c9dc9089361995085298598992bf69f11302a5dbd2c-chan#13696: Attempt to send HTLC but unowned (CHANNELD_NORMAL)
```
It's your node id. But I don't know, what is "Attempt to send HTLC but unowned" means.

Also, I've checked for `WIRE_CHANNEL_DISABLED` and I have some:
```
2021-05-28T20:32:20.990Z INFO    02f3069a342ae2883a6f29e275f06f28a56a6ea2e2d96f5888a3266444dcf542b6-chan#13597: htlc 14017 failed from 2th node with code 0x1014 (WIRE_CHANNEL_DISABLED)
...
2021-05-29T06:38:59.841Z INFO    024a8228d764091fce2ed67e1a7404f83e38ea3c7cb42030a2789e73cf3b341365-chan#13580: htlc 14104 failed from 2th node with code 0x1014 (WIRE_CHANNEL_DISABLED)
```
The second one surely comes from a c-lightning node.

Reply:
About the `msat_factor_max`, could it be a config option? I'm still not convinced that it is a good idea. I would like the option to set it to 1 if it isn't working well. Or may try different values too.

Reply:
Maybe I remove the msatoshi scaling all together since I'm also unsure. We need to to A-B tests to know if its better.

What is your opinion about the incremental approach to maxhops in trying short routes first? Do we want this?

Reply:
Other way would be to keep the msatoshi scaling in a modified way, maybe less aggressive and turning it to neutral (1) when it failed to a certain maxhop, so the algorithm still tries all the routes...

Reply:
I'm not sure about the incremental approach of `maxhops`. If we call `getroute(maxhops=4)` and after that `getroute(maxhops=5)` it will do partly the same route search twice?

Could this incremental search be configurable? So I may choose just one search with for example `maxhops=5`?

Reply:
> I've checked my logs and I've found this:
> 
> ```
> 2021-05-28T22:38:25.203Z INFO    02a4aae2e021b0785f2f2a7c9dc9089361995085298598992bf69f11302a5dbd2c-chan#13696: Attempt to send HTLC but unowned (CHANNELD_NORMAL)
> ```
> 
> It's your node id. But I don't know, what is "Attempt to send HTLC but unowned" means.
> ...

Talked to cdecker about this, my node happened to turn offline just at the moment your node tried to send me :D